### PR TITLE
Implement `extends` keyword

### LINKED ISSUE #36: Extends keyword
## Inheriting properties from an already defined file using `extends` keyword
In simple terms, allow modifying values of a file created using the `new` keyword without copying the whole file.  
This would be really useful for custom advancements that are different by a single line or the likes.  
```js
new advancement(blue_reward) {
  "display": {
    "icon": {
      "item": "minecraft:blue_concrete"
    },
    "title": "Example",
    "description": "example",
    "frame": "goal"
  },
  "criteria": {
    "requirement": {
      "trigger": "minecraft:tick",
      "conditions": {
        "player": {
          "team": "blue"
        }
      }
    }
  },
  "rewards": {
    "function": "example:reward"
  }
}

new advancement(red_reward) extends(blue_reward) {
  "display": {
    "icon": {
      "item": "minecraft:red_concrete"
    },
  },
  "criteria": {
    "requirement": {
      "conditions": {
        "player": {
          "team": "red"
        }
      }
    }
  }
}
```
This would allow one to copy and modify files that require minor changes without completely copying the contents of the file.  
This also allows making the same changes to the files faster for example if the `reward` were changed to also include experience, all the files extending the parent would get the change.

### PR DESCRIPTION
**Pull request by Nico314159**
Resolves #36

### CODE REVIEW

File: src/jmc/compile/utils.py [hunk 1]
```diff
@@ -248,3 +249,22 @@ def rand_java_int() -> int:
...
+def deep_merge(first: dict, second: dict) -> dict:
+    """
+    Recursively merges two dictionaries together.
+    (i.e. it also merges any dicts inside of them)
+
+    :param first: The dictionary to merge stuff to
+    :param second: The dictionary being merged into the first
+    :return: A new dict that is the merger of both inputs
+    """
+    print("\033[96;40m", f"first: {first}", "\033[0;0;40m")
```
Comment (WingedSeal) [inline src/jmc/compile/utils.py:262]: Is this supposed to be here?

File: src/jmc/compile/lexer.py [hunk 2]
```diff
@@ -461,10 +467,29 @@ def parse_new(self, tokenizer: Tokenizer,
...
+            while True: break
+            super_name = prefix + convention_jmc_to_mc(
+                command[4], tokenizer, is_make_lower=False, substr=(1, -1))
+            if namespace in Header().namespace_overrides:
+                super_path = namespace + "/" + json_type + \
+                    "/" + super_name[len(namespace) + 1:]
+            else:
+                super_path = json_type + "/" + super_name
+
+            try:
```
Comment (WingedSeal) [inline src/jmc/compile/lexer.py:485]: What's the point of try-except here?

File: src/jmc/compile/lexer.py [hunk 3]
```diff
@@ -461,10 +467,29 @@ def parse_new(self, tokenizer: Tokenizer,
...
+                super_json = self.datapack.jsons[super_path]
+            except KeyError:
+                raise KeyError()
+
+            assert isinstance (super_json, dict);
```
Comment (WingedSeal) [inline src/jmc/compile/lexer.py:490]: Please fix spacing

File: src/jmc/compile/utils.py [hunk 4]
```diff
@@ -248,3 +249,22 @@ def rand_java_int() -> int:
 
 def is_decorator(string: str) -> bool:
     return (len(string) > 2 and string.startswith("@"))
+
+def deep_merge(first: dict, second: dict) -> dict:
+    """
```
Comment (WingedSeal) [inline src/jmc/compile/utils.py:254]: Nice docs!

**Review by WingedSeal — CHANGES_REQUESTED** [review]

File: src/jmc/compile/utils.py (hunk 1 above)
Comment (Nico314159) [reply src/jmc/compile/utils.py:262]: Shoot I missed one when removing print statements

File: src/jmc/compile/lexer.py (hunk 2 above)
Comment (Nico314159) [reply src/jmc/compile/lexer.py:485]: Fixed

File: src/jmc/compile/lexer.py [hunk 5]
```diff
@@ -461,10 +467,31 @@ def parse_new(self, tokenizer: Tokenizer,
...
         except JSONDecodeError as error:
-            raise JMCDecodeJSONError(error, command[3], tokenizer) from error
+            raise JMCDecodeJSONError(error, command[-1], tokenizer) from error
         if not json:
             raise JMCSyntaxException(
-                "JSON content cannot be empty", command[3], tokenizer)
+                "JSON content cannot be empty", command[-1], tokenizer)
+
+        if has_extends_arg:
+            while True: break
```
Comment (WingedSeal) [inline src/jmc/compile/lexer.py:476]: ???

**Review by WingedSeal — CHANGES_REQUESTED** [review]

File: src/jmc/compile/lexer.py (hunk 5 above)
Comment (Nico314159) [reply src/jmc/compile/lexer.py:476]: I'm just as confused as you are

**Pull request merged**